### PR TITLE
Add RawDataContainer, SNPE eval and pass accept common dataloader, Migrate snpe examples to workflow jsons

### PR DESCRIPTION
**Pull request by jambayk**
## Describe your changes
This PR introduces a new data container called `RawDataContainer` where the data is saved in directory as numpy binary files. 
See the doc for `RawDataset` for more details. 
SNPE quantization and evaluation converts the generic dataloader to snpe dataloader so they are not limited snpe data loaders anymore.

In the future, we can expand the dataset to support diferent types of data (.npy, .npz, pickle, etc), 

The snpe examples have been changed to workflow config jsons that use the raw data config. 
## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Format your code by running `pre-commit run --all-files`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
SNPE evaluation and quantization now accept generic dataloaders such as torch dataloader. They no longer require users to provide SNPE dataloaders.

### CODE REVIEW

File: olive/engine/engine.py [hunk 1]
```diff
@@ -289,7 +289,7 @@ def run(
                     result_name = f"{prefix_output_name}metrics"
                     results_path = output_dir / f"{result_name}.json"
                     with open(results_path, "w") as f:
-                        f.write(results.json())
+                        json.dump(results.to_json(), f, indent=4)
```
Comment (guotuofeng) [inline olive/engine/engine.py:292]: >to_json [](http://example.com/codeflow?start=42&length=7)

what's the difference between to_json and json()?

File: olive/data/component/dataloader.py [hunk 2]
```diff
@@ -48,3 +55,120 @@ def rewind(self):
...
     return _CalibrationDataReader(dataloader, **kwargs)
+
+
+@Registry.register_dataloader()
+def default_snpe_dataloader(_dataset, snpe_model, batch_size=None):
+    from olive.snpe.data_loader import SNPEProcessedDataLoader
+    from olive.snpe.utils.input_list import create_input_list
+
+    class _SNPEProcessedDataLoader(SNPEProcessedDataLoader):
+        def __init__(self, dataset, snpe_model, batch_size=None):
```
Comment (guotuofeng) [inline olive/data/component/dataloader.py:66]: >snpe_model [](http://example.com/codeflow?start=36&length=10)

seems we only need the io_config?

File: olive/engine/engine.py (hunk 1 above)
Comment (jambayk) [reply olive/engine/engine.py:292]: the .json() returns a json string (it is a built-in method of pydantic BaseModel). `to_json` is a method of `ConfigBase` that returns a json as python dictionary.

File: olive/data/container/data_container.py [hunk 3]
```diff
@@ -65,6 +65,18 @@ def create_dataloader(self):
         pre_process_dataset = self.pre_process(dataset)
         return self.dataloader(pre_process_dataset)
 
+    def create_snpe_dataloader(self, model, evaluation=True):
```
Comment (guotuofeng) [inline olive/data/container/data_container.py:68]: >create_snpe_dataloader [](http://example.com/codeflow?start=8&length=22)

do we need add new signature?

Comment (jambayk) [reply olive/data/container/data_container.py:68]: I just discussed this with Jiapeng. The snpe dataloader is used only for snpe model evaluation or quantization, so it is better kept as a separate method. Unless, we want to add different condition to the `create_dataloader` and `create_calibration_dataloader` methods.

File: olive/data/component/dataloader.py (hunk 2 above)
Comment (jambayk) [reply olive/data/component/dataloader.py:66]: yes, it only needs the `input_names` and `input_shapes`.

File: olive/evaluator/olive_evaluator.py [hunk 4]
```diff
@@ -156,22 +156,23 @@ def get_user_config(metric: Metric):
         eval_func = user_module.load_object(evaluate_func)
```
[the remark this answers is not in the document]
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:156]: It depends on the model and data. 

for the user_config part, it is used to create dummy data. 
for snpe_model, we need to get it from the model itself since we don't know the model's io config before hand. SNPE permutes the input shapes from `batchXchannelXwidthXheight` to `batchXwidthXheightXchannel` so we need both the original shapes (from data) and the model shape.

[the remark this answers is not in the document]
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:156]: for onnx model, we cannot use it to create dummy data so the latency metric user config might have io config inside them.

[the remark this answers is not in the document]
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:156]: I think we could revisit these later if needed.

[the remark this answers is not in the document]
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:156]: @jiapeng I think there are some plans to maybe remove `user_config` method of creating dataloaders in the future? Maybe we can review this then. Since the user_config `input_names`, `output_names` way of creating dummy data is already covered by the `DummyDataContainer`.

File: olive/engine/engine.py (hunk 1 above)
Comment (guotuofeng) [reply olive/engine/engine.py:292]: could I suppose the json.load(json()) == to_json()?

File: olive/data/container/data_container.py (hunk 3 above)
Comment (guotuofeng) [reply olive/data/container/data_container.py:68]: Is it possible there would be lot of customized create_xx_dataloader in future? that's my concerns.

File: olive/engine/engine.py (hunk 1 above)
Comment (jambayk) [reply olive/engine/engine.py:292]: Yes, it only pretty prints the json file. the contents are the same.

File: olive/data/container/data_container.py (hunk 3 above)
Comment (jambayk) [reply olive/data/container/data_container.py:68]: That is a valid concern. I added this directly to the container since `SNPEModel` is an olive model type and it also needs snpe dataloader. I think for other passes, they would need to do the conversion from olive dataloader to their specific type in the pass itself. 

I could also do the same. remove this method and instead get olive dataloader. And the convert it to snpe dataloader in the snpe evaluation and quantization pass.

File: olive/data/component/dataloader.py (hunk 2 above)
Comment (guotuofeng) [reply olive/data/component/dataloader.py:66]: if so, should we pass io_config here instead of the mdoel itself?

File: olive/data/container/data_container.py (hunk 3 above)
Comment (trajepl) [reply olive/data/container/data_container.py:68]: How about this? @jambayk 
You can inherit the RawDataContainer and change the dataloader to snpe_dataloader. In this way, we can keep the base class as simple as possible. Then the usage in SNPE case will be clearer.
```python
@Registry.register(DataContainerType.DATA_CONTAINER)
class SNPEDataContainer(RawDataContainer):
default_components_type: ClassVar[dict] = {
    DataComponentType.LOAD_DATASET.value: "raw_dataset"
    DataComponentType.DATALOADER.value: "default_snpe_dataloader"
}
```

Comment (trajepl) [reply olive/data/container/data_container.py:68]: Or we can create snpe template in `https://github.com/microsoft/Olive/pull/333/files/f9b7c7eac227bfa2b9f6f05ff80516184b472c33#diff-bfa09a24344a0657b60de3f87a8b40be86b65f9e34ab3ac259f0c631ba9d6c3c` based on RawDataContainer.

File: olive/evaluator/olive_evaluator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (trajepl) [reply olive/evaluator/olive_evaluator.py:156]: Sounds good to me currently. Thanks.

In the future, maybe we can consider to let io_config spread to onnx model as well?
 Or another method for dynamic axis, we can assign the value if user does not provide it. Just make sure it is consist during the comparation in perf_tuning?

File: olive/data/component/dataloader.py (hunk 2 above)
Comment (jambayk) [reply olive/data/component/dataloader.py]: Done.

File: olive/evaluator/olive_evaluator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:156]: Auto assigning would work too but I don't know if the perf tuning results would be as useful as using user provided shapes. Since we want to tune for data the user expects to use the model irl.

File: olive/data/container/data_container.py (hunk 3 above)
Comment (trajepl) [reply olive/data/container/data_container.py]: A bit differences with offline discussion where I were thinking to replace the default dataloader to `default_snpe_dataloader` in `RawDataContainer`. Then the `RawDataContainer` only serves SNPE. I agreed that is not the way to go.

[1] The new suggestion is that we keep the `RawDataContainer` and to produce the `SNPEDataContainer`.  Then in SNPE config, we can directly use `SNPEDataContainer`. But for onnx/torch model, we can use `RawDataContainer`. **That is determined by users.**

[2] For this, it sounds good to me as well. 
> If people are worried about the new create_snpe_dataloader method, I think a better solution is to not add the method. The evaluation and quantization code will then do the following:
datacontainer.create_dataloader -> regular dataloader -> Convert regular dataloader to snpe dataloader


For me, I have no strong tendency on method [1] or [2]. Both of them can resolve Mike's concern.

Comment (trajepl) [reply olive/data/container/data_container.py]: Besides, quick question about this: I do not get your points. Is there any necessary case where we must call snpe_dataloader independently? From my understanding, all the method in data components can be call independently.
>  The snpe dataloader code has no dependency on the RawDataContainer. It can work with any datacontainer.

File: olive/evaluator/olive_evaluator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (trajepl) [reply olive/evaluator/olive_evaluator.py:156]: Good catch. The real case will be more flexible. The auto io_config assigning will work for perf_tuning accurately only when user know we will use the io_config in input model for perf tuning as well.

[the remark this answers is not in the document]
Comment (jambayk) [reply olive/evaluator/olive_evaluator.py:156]: Okay. I will close this conversation for now since it is related to a different feature.

File: olive/data/container/data_container.py (hunk 3 above)
Comment (jambayk) [reply olive/data/container/data_container.py]: actually, pushed the changes for [2] directly. Please let me know if that addresses the concerns.

File: olive/data/template.py [hunk 5]
```diff
@@ -55,3 +55,33 @@ def huggingface_data_config_template(model_name, task, **kwargs) -> DataConfig:
             **kwargs,
         },
     )
+
+
+def raw_data_config_template(
```
[the remark this answers is not in the document]
Comment (trajepl) [reply olive/data/template.py:60]: It is fine to keep it for me. I created the template module for:
1. demonstrate popular data config.
2. simplify the call action.

**Review by trajepl — APPROVED** [review]
